llvm/tools/sycl-post-link/ModuleSplitter.cpp

```diff
@@ -307,13 +315,25 @@ class CallGraph {
                ? make_range(EmptySet.begin(), EmptySet.end())
                : make_range(It->second.begin(), It->second.end());
   }
+
+  iterator_range<FunctionSet::const_iterator> addrTakenFunctions() const {
+    return make_range(AddrTakenFunctions.begin(), AddrTakenFunctions.end());
+  }
 };
 
 void collectFunctionsToExtract(SetVector<const GlobalValue *> &GVs,
                                const EntryPointGroup &ModuleEntryPoints,
                                const CallGraph &Deps) {
   for (const auto *F : ModuleEntryPoints.Functions)
     GVs.insert(F);
+  // It is conservatively assumed that any address-taken function can be invoked
+  // or otherwise used by any function in any module split from the initial one.
+  // So such functions along with the call graphs they start are always
+  // extracted (and duplicated in each split module).
```

Comment:
Can we move such function in a separate module and link only with modules, which use indirect invocation?

Reply:
That could be done. But maybe as a separate (big) effort. There can potentially be other uses besides indirect invocation which would require availability of an address taken function (e.g. address comparison) and some data flow analysis might be needed. Conservative approach should be good enough to start with, I believe.

Reply:
> That could be done. But maybe as a separate (big) effort. There can potentially be other uses besides indirect invocation which would require availability of an address taken function (e.g. address comparison) and some data flow analysis might be needed. Conservative approach should be good enough to start with, I believe.

A TODO could be added for this. Thanks

Reply:
Done